### PR TITLE
[Feature] Add Windows CE support

### PR DESCRIPTION
**Pull request by e-i-n-s**
Changelog: Feature: Add Windows CE support
Docs: TBD

A small background of this PR:
Our company (a JFrog Artifactory customer) still needs some legacy builds for the Windows CE operating system. A native support of conan for Windows CE makes our conanfiles.py much more readable. Since CMake offers a good support of Windows CE, it would be nice to support this platform natively in Conan as well.

- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

I'll create the documentation as soon as I have a commitment from you to merge this PR.

### CODE REVIEW

**Comment by e-i-n-s** [comment]
@lasote
Please let me know if you need any additional information.
Best regards,
Marco

**Comment by lasote** [comment]
Hi Marco,
I'll open soon a review asking some question. I think we can move this forward for the next release.
Thanks for the contribution

File: conans/client/tools/win.py [hunk 1]
```diff
@@ -364,36 +364,10 @@ def vcvars_command(settings, arch=None, compiler_version=None, force=False, vcva
         last_version = latest_vs_version_installed(output=output)
 
         compiler_version = compiler_version or last_version
-    os_setting = settings.get_safe("os")
     if not compiler_version:
         raise ConanException("compiler.version setting required for vcvars not defined")
 
-    # https://msdn.microsoft.com/en-us/library/f2ccy3wt.aspx
```
Comment (lasote) [inline conans/client/tools/win.py]: Why this movement of code? Please, revert it, it is hard to see what you changed in the diff and we would like to see only the "effective" code.

Comment (lasote) [reply conans/client/tools/win.py]: Maybe you can just introduce an "if" at the beginning and returning.

**Comment by e-i-n-s** [comment]
@lasote thank you very much for the frist review.

I think it makes sense to discuss first the "platform: ANY" in the settings.yml.

In Windows CE there are some predefined "platforms" (like for Windows Mobile). But when using customized hardware, there's the need of creating an own platform (using the "Platform Builder for Windows CE").
Some examples for the platform name:
- **SDK_AM335X_SK_WEC2013_V310** (Source: https://cmake.org/cmake/help/v3.14/manual/cmake-toolchains.7.html#cross-compiling-for-windows-ce)
- **STANDARDSDK_500 (ARMV4I)**
- **Windows Mobile 5.0 Pocket PC SDK (ARMV4I)**
- **Toradex_CE800 (ARMV7)**
- **YOUR_SDK_NAME (ARMv4I)**
- **YOUR_SDK_NAME**

Example of an CMake generator expression: "Visual Studio 9 2008 YOUR_SDK_NAME (ARMv4I)".

Because any values are possible (except empty), the value ANY is necessary.

I could change "platform" to "sdk", is this ok for you?

**Comment by lasote** [comment]
What about the `/p:Platform` flag in the `MSBuild` command? do you have some docs about it?
For CMake, you might be good using `CONAN_CMAKE_GENERATOR_PLATFORM`. Being the platform open to any value it loses a bit the sense of a setting, so we want to make sure if we want to model that as a setting.

File: conans/client/build/cmake_flags.py [hunk 2]
```diff
@@ -33,7 +33,7 @@ def get_generator(settings):
     compiler = settings.get_safe("compiler")
     arch = settings.get_safe("arch")
     compiler_version = settings.get_safe("compiler.version")
-    os_build, _, _, _ = get_cross_building_settings(settings)
+    os_build, _, os_host, _ = get_cross_building_settings(settings)
```
[the remark this answers is not in the document]
Comment (e-i-n-s) [reply conans/client/build/cmake_flags.py]: `os` is used by `import os`, I will use `os_host`, other proposals? (https://docs.conan.io/en/latest/systems_cross_building/cross_building.html)

File: conans/client/build/cmake_flags.py [hunk 3]
```diff
@@ -55,7 +55,7 @@ def get_generator(settings):
         if Version(compiler_version) < "16":
             if arch == "x86_64":
                 base += " Win64"
-            elif "arm" in arch:
+            elif os_host != "WindowsCE" and "arm" in arch:
```
[the remark this answers is not in the document]
Comment (e-i-n-s) [reply conans/client/build/cmake_flags.py]: Windows CE (usually ARM, sometimes x86 and others) is an exception for Visual Studio...

`Visual Studio 15 2017 ARM` is cross-compiling for `Windows 10 on ARM`. In Windows CE the architecture is defined in the "platform SDK" and therefore not relevant in the CMake call.

I will add a check for `Windows` os and remove the check for `WindowsCE`.

File: conans/client/tools/win.py (hunk 1 above)
Comment (e-i-n-s) [reply conans/client/tools/win.py]: I extracted the code to make it testable for visual studio versions, which are not installed in the CI build infrastructure...
I adapted the unit test and reverted the extracting.

**Comment by e-i-n-s** [comment]
@lasote 
There's no separate Windows CE documentation for the  `/p:Platform` flag in the `MSBuild` command. In MSVC the Windows CE platforms are handled like `Win32` or `x64`. For instance the `Platform` could be `YOUR_SDK_NAME (ARMv4I)` or `ANOTHER_SDK_NAME`.

The `CONAN_CMAKE_GENERATOR_PLATFORM` is not really an option, because every Windows CE platform has different binaries. In the end we need conan to get different packages/binaries for each Windows CE platform we are using.

Let me know if you need any additional information. I updated the code, please review.

File: conans/client/build/cmake_flags.py [hunk 4]
```diff
@@ -52,7 +53,7 @@ def get_generator(settings):
                     '16': '16 2019'}
         base = "Visual Studio %s" % _visuals.get(compiler_version,
                                                  "UnknownVersion %s" % compiler_version)
-        if Version(compiler_version) < "16":
+        if os_host == "Windows" and Version(compiler_version) < "16":
```
Comment (lasote) [inline conans/client/build/cmake_flags.py]: I don't know if "WindowsStore" would be affected by this if. Should we use `if os_host != "WindowsCE" and ...` ?

Comment (e-i-n-s) [reply conans/client/build/cmake_flags.py]: According to this CMake tests the cmake generator doesn't include "ARM" or "Win64" suffix:
https://gitlab.kitware.com/cmake/cmake/commit/22282d69316eaf6e4a2b50af2ca7585a58f00d28

I'll test "WindowsStore" and, if necessary, request a separate PR.

**Comment by memsharded** [comment]
Hi,

It would be great if you could update the docs with a PR, for example here: https://docs.conan.io/en/latest/reference/config_files/settings.yml.html

And then add the PR link in the first comment of this PR, where ``Docs: TBD``, as that is used to automatically generate the changelog.

Many thanks again @e-i-n-s !!

**Pull request merged**